src/aws/flb_aws_util.c

```diff
@@ -507,3 +512,167 @@ int flb_read_file(const char *path, char **out_buf, size_t *out_size)
 
     return 0;
 }
+
+/* Generic replace function for strings. */
+char* replace_uri_tokens(const char* original_string, const char* current_word, 
+                         const char* new_word) 
+{ 
+    char *result; 
+    int i = 0;
+    int count = 0; 
+    int new_word_len = strlen(new_word); 
+    int old_word_len = strlen(current_word); 
+  
+    for (i = 0; original_string[i] != '\0'; i++) { 
+        if (strstr(&original_string[i], current_word) == &original_string[i]) { 
+            count++; 
+            i += old_word_len - 1; 
+        } 
+    }
+
+    result = flb_sds_create_size(i + count * (new_word_len - old_word_len) + 1);
+    if (!result) {
+        flb_errno();
+        return NULL;
+    }
+
+    i = 0; 
+    while (*original_string) { 
+        if (strstr(original_string, current_word) == original_string) { 
+            strcpy(&result[i], new_word); 
+            i += new_word_len; 
+            original_string += old_word_len; 
+        } 
+        else
+            result[i++] = *original_string++; 
+    } 
+
+    result[i] = '\0'; 
+    return result; 
+} 
+
+/* Constructs S3 object key as per the format. */
+flb_sds_t get_s3_key(const char *format, time_t time, char *tag, char *tag_delimiter)
+{
+    int i = 0;
+    int ret = 0;
+    char *tag_token = NULL;
+    flb_sds_t tmp = NULL;
+    flb_sds_t buf = NULL;
+    flb_sds_t s3_key = NULL;
+    flb_sds_t tmp_key = NULL;
+    flb_sds_t tmp_tag = NULL;
+    struct tm *gmt = NULL;
+    
+    if (strlen(format) > S3_KEY_SIZE){
+        flb_warn("[s3_key] Object key length is longer than the 1024 character limit.");
+    }
+
+    tmp_tag = flb_sds_create_len(tag, strlen(tag) + 1);
+    if(!tmp_tag){
+        flb_errno();
+        return NULL;
+    }
+
+    s3_key = flb_sds_create_len(format, S3_KEY_SIZE);
```

Comment:
Go and look at the code for flb_sds_create_len in flb_sds.c.

The 2nd arg, the len arg, is the amount of memory to copy from the first arg. So the 2nd arg should be the length of format.

Reply:
The intuition here was to set s3_key with the max length 1024 because of 2 reasons

- max length of s3_key cannot exceed 1024 

- the function parameter format can be greater than 1024 in which case only the first 1024 is copied.

Now I see how flb_sds_create_len uses memcpy internally which I learnt is not safe. Thanks for pointing this out!! 

We can switch this to strlen(format) and take it in even if it exceeds 1024 because my code does time formatting at the end where strftime copies a max of 1024 only. 
```
    ret = strftime(tmp, S3_KEY_SIZE, s3_key, gmt);
    
    if(ret == 0){
        flb_warn("[s3_key] Object key length is longer than the 1024 character limit.");
    }

    flb_sds_destroy(s3_key);
    s3_key = tmp;
```
This will ensure we never send any key greater than 1024.